### PR TITLE
[1466] Citizen Project validation

### PR DESCRIPTION
**Pull request by ottaviano**

### CODE REVIEW

File: app/config/services/twig.xml [hunk 1]
```diff
@@ -54,6 +54,10 @@
 
         <service id="AppBundle\Twig\MarkdownExtension" />
 
+        <service id="AppBundle\Twig\ReadMoreExtension">
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply app/config/services/twig.xml]: @B-Galati fixed 👍

File: src/Committee/CommitteeManager.php [hunk 2]
```diff
@@ -457,7 +457,7 @@ public function changePrivilege(Adherent $adherent, Committee $committee, string
         }
     }
 
-    public function getCoordinatorCommittees(Adherent $coordinator, CommitteeFilters &$filters): array
+    public function getCoordinatorCommittees(Adherent $coordinator, CommitteeFilter $filters): array
```
Comment (tony-tran) [inline src/Committee/CommitteeManager.php]: Maybe you can change the name of variable too ?

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * @Route("/espace-coordinateur/projet-citoyen")
+ * @Security("has_role('ROLE_COORDINATOR_CITIZEN_PROJECT')")
+ */
+class CoordinatorCitizenProjectController extends Controller
+{
+    /**
+     * @Route(path="/list", name="app_coordinator_citizen_project")
+     */
+    public function listAction(Request $request)
```
Comment (tony-tran) [inline src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: Missing return typehint or remove it for `validateAction()`

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            } catch (BaseGroupException $exception) {
+                throw $this->createNotFoundException(sprintf('Citizen project %u has already been treated by an administrator.', $project->getId()), $exception);
+            }
+        } else {
+            foreach ($form->getErrors(true) as $error) {
+                $this->addFlash('error_'.$form->getData()->getId(), $error->getMessage());
+            }
+        }
+
+        return $this->redirectToRoute('app_coordinator_citizen_project', ['s' => CitizenProject::PENDING]);
```
Comment (tony-tran) [inline src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: For `app_coordinator_citizen_project ` route, there is no `s` parameter ?

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+
+/**
+ * @Route("/espace-coordinateur/projet-citoyen")
+ * @Security("has_role('ROLE_COORDINATOR_CITIZEN_PROJECT')")
+ */
+class CoordinatorCitizenProjectController extends Controller
+{
+    /**
+     * @Route(path="/list", name="app_coordinator_citizen_project")
```
Comment (tony-tran) [inline src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: Please add `@Method('GET')`

File: app/config/security.yml [hunk 6]
```diff
@@ -261,6 +266,7 @@ security:
         - { path: '^/espace-adherent/documents', roles: [ROLE_ADHERENT] }
         - { path: '^/espace-adherent/contacter', roles: [ROLE_ADHERENT] }
         - { path: '^/espace-adherent/mes-evenements', roles: [ROLE_ADHERENT] }
+        - { path: '^/espace-coordinateur/.*', roles: [ROLE_COORDINATOR] }
```
Comment (Remg) [inline app/config/security.yml]: wouldn't `path: '^/espace-coordinateur'` be enough ?

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php (hunk 4 above)
Comment (Remg) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: `s` (for _status_) parameter is used for filter purposes;
see https://github.com/EnMarche/en-marche.fr/pull/1825/files#diff-732435785991ed33606c03f60da392d7R29
& https://github.com/EnMarche/en-marche.fr/pull/1825/files#diff-ac23d0cd9617d08e52f5b3dd2f50fc8aR59

File: src/Entity/Adherent.php [hunk 7]
```diff
@@ -148,11 +149,11 @@ class Adherent implements UserInterface, GeoPointInterface, EncoderAwareInterfac
     private $managedArea;
 
     /**
-     * @var CoordinatorManagedArea|null
+     * @var CoordinatorManagedArea[]|Collection|null
      *
-     * @ORM\OneToOne(targetEntity="AppBundle\Entity\CoordinatorManagedArea", mappedBy="adherent", cascade={"persist"})
+     * @ORM\OneToMany(targetEntity="AppBundle\Entity\CoordinatorManagedArea", mappedBy="adherent", cascade={"persist", "remove"}, orphanRemoval=true)
```
Comment (Remg) [inline src/Entity/Adherent.php]: Since a user can't be a coordinator of both `CitizenProject` and `Comittee` sectors, why is this change needed ?

Comment (ottaviano) [reply src/Entity/Adherent.php]: @Remg, he must be able in code to be a coordinator of both sectors: cp and committee, but today we must allow only one choice (cp OR committee sector) in admin panel. In the future, we have just to activate `allow_add` & `allow_delete` options on admin form to allow to set many sectors and that's all.

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php (hunk 5 above)
Comment (ottaviano) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: done 👍

File: app/config/security.yml (hunk 6 above)
Comment (ottaviano) [reply app/config/security.yml]: yes, done 👍

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php (hunk 4 above)
Comment (ottaviano) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: @tony-tran yes, it's a query param (GET) and not of route. 
@Remg fixed 👍

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php (hunk 3 above)
Comment (ottaviano) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: done 👍

File: src/Committee/CommitteeManager.php (hunk 2 above)
Comment (ottaviano) [reply src/Committee/CommitteeManager.php]: it's an oversight, done 👍

**Comment by ottaviano** [comment]
I'll rebase and squash at the end of review and I'll try to add some test today.

File: src/Form/EventListener/CoordinatorManagedAreaListener.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        $form = $event->getForm();
+        /** @var Adherent $adherent */
+        $adherent = $event->getData();
+
+        if (!$adherent instanceof Adherent) {
+            return;
+        }
+
+        if (0 === $adherent->getCoordinatorManagedAreas()->count()) {
+            $form->get('coordinatorManagedAreas')->setData(
```
Comment (HeahDude) [inline src/Form/EventListener/CoordinatorManagedAreaListener.php]: Should be done above in PRE_SET_DATA with a chained call to `setDataLocked(true)` otherwise the form tree is not in synch.

File: templates/coordinator/_citizen_project_list.html.twig [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        </tr>
+    </thead>
+    <tbody id="citizen-projects-list">
+        {% for project in results %}
+            {% if not project.isApproved %}
+                {% set form = forms[project.id] %}
+
+                {{ form_start(form) }}
+
+                {{ form_errors(form) }}
```
Comment (HeahDude) [inline templates/coordinator/_citizen_project_list.html.twig]: Please indent the code between `form_start` and `form_end` as you would with regular form tags.

File: translations/messages.fr.yml [hunk 10]
```diff
@@ -245,3 +248,11 @@ citizen_project.committees: Comité en soutien|Comités en soutien"
 # Errors
 #
 recaptcha.error: Une erreur s'est produite, pouvez-vous réessayer ?
+
+#
+# Coordinator
+#
+coordinator.sector.committee: Secteur - Comités
+coordinator.sector.citizen_project: Secteur - Projets citoyens
+coordinator.label.codes: Codes des zones gérés
```
Comment (HeahDude) [inline translations/messages.fr.yml]: gérées

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Security;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+
+/**
+ * @Route("/espace-coordinateur/projet-citoyen")
+ * @Security("has_role('ROLE_COORDINATOR_CITIZEN_PROJECT')")
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: yes, like any other role, but it wasn't planned :)

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        try {
+            $filters = CitizenProjectFilter::fromQueryString($request);
+        } catch (\UnexpectedValueException $e) {
+            throw new BadRequestHttpException('Unexpected Citizen Project status in the query string.', $e);
+        }
+
+        $results = $this->get('app.citizen_project.manager')->getCoordinatorCitizenProjects($this->getUser(), $filters);
+
+        $forms = [];
+        array_walk($results, function (CitizenProject $project) use (&$forms, $filters) {
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: fixed 👍

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            ->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            try {
+                if ($form->get('refuse')->isClicked()) {
+                    $this->get('app.citizen_project.authority')->preRefuse($project);
+                    $this->addFlash('info', 'Merci. Votre appréciation a été transmise à nos équipes.');
+                }
+
+                if ($form->get('accept')->isClicked()) {
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: done

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+                    $this->get('app.citizen_project.authority')->preApprove($project);
+                    $this->addFlash('info', 'Merci. Votre appréciation a été transmise à nos équipes.');
+                }
+            } catch (BaseGroupException $exception) {
+                throw $this->createNotFoundException(sprintf('Citizen project %u has already been treated by an administrator.', $project->getId()), $exception);
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: ok, fixed by flash info

File: src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php [hunk 15]
```diff
@@ -1,27 +1,52 @@
...
+abstract class AbstractCoordinatorAreaFilter
 {
-    const PER_PAGE = 20;
+    public const PER_PAGE = 20;
 
-    const PARAMETER_OFFSET = 'o';
-    const PARAMETER_PAGE = 'p';
-    const PARAMETER_STATUS = 's';
+    private const PARAMETER_OFFSET = 'o';
+    public const PARAMETER_STATUS = 's';
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php]: done

File: src/Controller/EnMarche/Coordinator/CoordinatorCommitteeController.php [hunk 16]
```diff
@@ -15,19 +15,19 @@
...
     /**
-     * @Route("/comites", name="app_coordinator_committees")
+     * @Route("/list", name="app_coordinator_committees")
      * @Method("GET")
      */
     public function committeesAction(Request $request): Response
     {
         try {
-            $filters = CommitteeFilters::fromQueryString($request);
+            $filters = CommitteeFilter::fromQueryString($request);
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Controller/EnMarche/Coordinator/CoordinatorCommitteeController.php]: done

File: src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php [hunk 17]
```diff
@@ -1,27 +1,52 @@
...
 
     private $status;
-    private $offset = 0;
+    protected $offset = 0;
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php]: done

File: src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php [hunk 18]
```diff
@@ -102,10 +104,34 @@ public function setOffset(int $offset): void
...
+    {
+        return $this->coordinator;
+    }
+
+    public function setCoordinator(Adherent $coordinator): void
+    {
+        $this->coordinator = $coordinator;
+    }
+
+    private function updateCount(QueryBuilder $qb, string $alias): void
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php]: done

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (HeahDude) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: It's important to know, if it can be then `is_granted` should be used instead of `has_role` to trigger the hierarchy voter and not rely on a simple token check.

File: src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php [hunk 19]
```diff
@@ -102,10 +104,34 @@ public function setOffset(int $offset): void
...
+            ->getSingleScalarResult();
+
+        $this->setCount($count);
+    }
+
     public function getQueryStringForOffset(?int $offset): string
     {
         $parameters = $this->getQueryStringParameters();
-        $parameters[self::PARAMETER_OFFSET] = $offset ?: $this->offset;
+        $parameters[self::PARAMETER_OFFSET] = $offset ?? $this->offset;
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php]: why? it's a bug fix for allow `0` offset.

File: src/Entity/Adherent.php [hunk 20]
```diff
@@ -965,4 +956,36 @@ public function removeTag(AdherentTag $adherentTag): void
...
+    {
+        if (!$this->coordinatorManagedAreas->contains($area)) {
+            $area->setAdherent($this);
+            $this->coordinatorManagedAreas->add($area);
+        }
+    }
+
+    public function removeCoordinatorManagedArea(CoordinatorManagedArea $area): void
+    {
+        if ($this->coordinatorManagedAreas->contains($area)) {
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Entity/Adherent.php]: ok, removed

File: src/Entity/Adherent.php [hunk 21]
```diff
@@ -965,4 +956,36 @@ public function removeTag(AdherentTag $adherentTag): void
...
+            $this->coordinatorManagedAreas->removeElement($area);
+        }
+    }
+
+    public function getCoordinatorManagedAreaCodesAsString(): string
+    {
+        return implode(', ', $this->coordinatorManagedAreas->map(function (CoordinatorManagedArea $area) {
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Entity/Adherent.php]: good idea, done

File: src/Entity/CoordinatorAreaTrait.php [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace AppBundle\Entity;
+
+use AppBundle\Exception\CoordinatorAreaAlreadyTreatedException;
+use Doctrine\ORM\Mapping as ORM;
+
+trait CoordinatorAreaTrait
+{
+    /**
+     * @var string
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Entity/CoordinatorAreaTrait.php]: done

File: src/Entity/CoordinatorAreaTrait.php [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * @ORM\Column(type="text", nullable=true)
+     */
+    private $coordinatorComment;
+
+    /**
+     * @var Adherent
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Entity/CoordinatorAreaTrait.php]: done

File: src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php (hunk 19 above)
[the remark this answers is not in the document]
Comment (HeahDude) [reply src/Coordinator/Filter/AbstractCoordinatorAreaFilter.php]: Ah ok ^^'

File: src/Entity/CoordinatorAreaTrait.php [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+        $this->status = self::PRE_REFUSED;
+    }
+
+    public function preApproved(): void
+    {
+        if ($this->isApproved() || $this->isRefused()) {
+            throw new CoordinatorAreaAlreadyTreatedException($this->uuid);
+        }
+
+        $this->status = self::PRE_APPROVED;
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Entity/CoordinatorAreaTrait.php]: yes, it's declared in BaseGroup entity, I can resolve the problem with abstraction

File: src/Form/EventListener/CoordinatorManagedAreaListener.php (hunk 8 above)
Comment (ottaviano) [reply src/Form/EventListener/CoordinatorManagedAreaListener.php]: done, but without the call setDataLocked, I didn't find this method in already builded form.

File: translations/messages.fr.yml (hunk 10 above)
Comment (ottaviano) [reply translations/messages.fr.yml]: done

File: src/Form/EventListener/CoordinatorManagedAreaListener.php (hunk 8 above)
Comment (HeahDude) [reply src/Form/EventListener/CoordinatorManagedAreaListener.php]: True, the listener has a form and not a builder, my bad, you did well 👍

File: src/Entity/Adherent.php [hunk 25]
```diff
@@ -148,11 +149,11 @@ class Adherent implements UserInterface, GeoPointInterface, EncoderAwareInterfac
     private $managedArea;
 
     /**
-     * @var CoordinatorManagedArea|null
+     * @var CoordinatorManagedArea[]|Collection|null
```
Comment (HeahDude) [inline src/Entity/Adherent.php]: Null must be removed.

File: src/Entity/CoordinatorManagedArea.php [hunk 26]
```diff
@@ -45,7 +58,7 @@ public function getAdherent(): ?Adherent
         return $this->adherent;
     }
 
-    public function setAdherent(Adherent $adherent = null)
+    public function setAdherent(Adherent $adherent = null): void
```
Comment (HeahDude) [inline src/Entity/CoordinatorManagedArea.php]: `public function setAdherent(?Adherent $adherent): void`

File: src/Twig/ReadMoreRuntime.php [hunk 27]
```diff
@@ -0,0 +1,48 @@
+<?php
+
+namespace AppBundle\Twig;
+
+use Twig\Extension\AbstractExtension;
+use Twig\TwigFilter;
+
+class ReadMoreRuntime extends AbstractExtension
```
Comment (HeahDude) [inline src/Twig/ReadMoreRuntime.php]: This is not a runtime but an extension, should either be renamed or decoupled.

File: src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (Remg) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: Actually, neither the role `ROLE_COORDINATOR_COMMITTEE` nor the role `ROLE_COORDINATOR_CITIZEN_PROJECT` can be inherited.

Although i see no valid reason to say they won't be inherited someday.

Let's go for `is_granted` ?

[the remark this answers is not in the document]
Comment (HeahDude) [reply src/Controller/EnMarche/Coordinator/CoordinatorCitizenProjectController.php]: I would say so.

**Comment by HeahDude** [comment]
It took time, but here we go, this is in now. Thank you very much @ottaviano.

**Comment by ottaviano** [comment]
GJ @HeahDude, tnx
